### PR TITLE
Bug 5449: Ignore SP and HTAB chars after chunk-size

### PR DESCRIPTION
**Pull request by uhliarik**
Prior to 2023 commit 951013d0, Squid accepted Transfer-Encoding chunks
with chunk-size followed by spaces or tabs (before CRLF). This HTTP
syntax violation was allowed to address Bug 4492 (fixed in 2017 commit
26f0a359). This change restores that fix functionality. FWIW, our
research shows that nginx and httpd also accept similar input.

### CODE REVIEW

**Comment by uhliarik** [comment]
Bug related to this MR - https://bugs.squid-cache.org/show_bug.cgi?id=5449

**Comment by uhliarik** [comment]
The only thing I'm not sure about this change is, if it is safe to expect that there can be unlimited amount of whitespaces. nginx skips only one '\t' or ' ' and apache httpd allow only 10 occurrences.

File: src/http/one/TeChunkedParser.cc
Comment (rousskov) [inline src/http/one/TeChunkedParser.cc:128]: The proposed addition effectively disables BWS handling logic inside ParseBws() called from parseChunkExtensions() below. It also mishandles partial/incomplete input that ends exactly at chunk-size (Squid should not fall through to the next line in that case but does).

If we decide to allow these spaces, both bugs should be eliminated. Here is an untested suggestion:

```suggestion
        ParseBws(tok); // some servers send "chunk-size BWS CRLF"
```

The suggestion above also adjusts the comment to avoid implying that this code was added for all "SP/TAB after chunk-size" cases. Some of those cases are actually already covered by BWS handling in parseChunkExtension() call below.

File: src/http/one/TeChunkedParser.cc
Comment (rousskov) [inline src/http/one/TeChunkedParser.cc:129]: The proposed change results in two sequential ParseBws() (or equivalent) calls because parseChunkExtensions() (correctly!) starts with ParseBws() as well. This problem alone is not a showstopper IMO, but we should keep this added complexity and waste of CPU cycles in mind when deciding whether to merge this PR.

It is probably possible to refector code to remove this waste, and it would be fair to require this PR to perform that refactoring, especially if it does not require heroic efforts.

**Review by rousskov — CHANGES_REQUESTED** [review]
I adjusted PR title to refer to the corresponding bug report. I adjusted PR description (i.e. future official commit message body) primarily to fix formatting and to add an XXX that we should try to resolve:

> PR description: Replies with such whitespace chars are pretty common (XXX: missing evidence).

@uhliarik, would it be possible to support the above assertion? The problem is clearly not _very_ "common" because it took a year for it to be reported despite the problematic code being present in v5+ releases. The bug report [cites](https://bugs.squid-cache.org/show_bug.cgi?id=5449#c0) what looks like some one-off CGI script, but I suspect you know of some more important/popular use cases that need Squid to be lenient here. Needless to say, Squid should not accept malformed message framing in attack-prone area just to preserve some obscure CGI script bugs or mimic questionable behavior by other proxies...

FWIW, if sufficient evidence of "pretty common" use is provided, I would be inclined to accept a fixed version (see specific change requests) of this PR.

**Comment by rousskov** [comment]
> PR description: Other proxies (e.g., nginx and httpd) are accepting them.

FWIW, AFAICT, httpd stated reason for accepting them is invalid -- they [cite](https://github.com/apache/httpd/blob/c717758efb3ea1a20c4aea6a7d64d3994a14a2ff/modules/http/http_filters.c#L192-L195) RFC [errata](https://www.rfc-editor.org/errata_search.php?rfc=7230&eid=4667) that does _not_ allow these chars when there is no chunk-ext! That errata resulted in the corresponding changes to subsequent RFCs, but those changes still do not allow `chunk-size SP CRLF` sequences.

**Comment by uhliarik** [comment]
> > PR description: Other proxies (e.g., nginx and httpd) are accepting them.
> 
> FWIW, AFAICT, httpd stated reason for accepting them is invalid -- they [cite](https://github.com/apache/httpd/blob/c717758efb3ea1a20c4aea6a7d64d3994a14a2ff/modules/http/http_filters.c#L192-L195) RFC [errata](https://www.rfc-editor.org/errata_search.php?rfc=7230&eid=4667) that does _not_ allow these chars when there is no chunk-ext! That errata resulted in the corresponding changes to subsequent RFCs, but those changes still do not allow `chunk-size SP CRLF` sequences.

Yes, I'm aware of that :(. When I was reading that comment and RFC Errata (And I actually saw that the change was proposed by you), I realized that this is related to chunk-ext and not to removing extra white chars. Obviously the comment does not describe the logic of the code, anyway it does not change anything on the fact that httpd skips those whitespaces. But you are right the comment is misleading.

**Comment by rousskov** [comment]
> The only thing I'm not sure about this change is, if it is safe to expect that there can be unlimited amount of whitespaces.

It is safe to reject any amount of prohibited whitespace. Everything else is unsafe. This PR is trading safety for compatibility. If we allow any space at all, then I would err on the side of code simplicity/reuse -- reject whatever we can (correctly) reject using existing code. I posted a specific [suggestion](https://github.com/squid-cache/squid/pull/1914#discussion_r1795888509) that was meant to do that.

File: src/http/one/TeChunkedParser.cc
Comment (uhliarik) [reply src/http/one/TeChunkedParser.cc:128]: That was the original change I made locally, but with this change squid didn't even send the first chunk of data. 

Squid was failing on the 

```
        tok.skipRequired("CRLF after [chunk-ext]", Http1::CrLf());
```
in Http::One::TeChunkedParser::parseChunkMetadataSuffix(Tokenizer &tok);


```
2024/10/10 10:55:41.162 kid1| 11,2| http.cc(1466) decodeAndWriteReplyBody: de-chunking failure: cannot skip CRLF after [chunk-ext]
    exception location: Tokenizer.cc(159) skipRequired
```

Can't it be caused because `ParseBws(tok);` also removes CR char?

Comment (rousskov) [reply src/http/one/TeChunkedParser.cc:128]: Thank you for testing my broken suggestion even before I made it!

I now see why my suggestion was bad: When relaxed_header_parser is on, One::Parser::WhitespaceCharacters() includes CR. In this parsing context, skipping CR prevents subsequent `tok.skipRequired(...CrLf())` from succeeding. The existing chunk-ext parser code does not suffer from this problem because its BWS sequences are always followed by something (e.g., semicolon or extension name) _before_ we get to CRLF.

Unfortunately, this change request is still valid, even if the suggested simple solution is not. We will need to write more code to fix the problems this change request identifies. In summary, there will be two types of BWS -- one that relaxed_header_parser can extend and one that always contains just SP and HTAB.

Do you mind if I push a change to your PR branch that adds support for those two types of BWS?

Comment (uhliarik) [reply src/http/one/TeChunkedParser.cc:128]: Yes, of course! Feel free to push it.

Yeah that's what I thought. It removed CR as well from the buffer and then it was not there when required for CrLf end of the chunk line.

**Comment by uhliarik** [comment]
I came up with some other changes after our discussion. I added one argument to `ParseBWS()` function which would either force skipping CharacterSet::WSP if set to true or by default (if the function argument is not set) it would skip Parser::WhitespaceCharacters() which is dependent on relaxed parsing option.

I also tried to optimize number of ParseBWS calls. In the new commit, `ParseBWS()` should be called only if there was no chunk-ext found, since if chunk-ext is found then the leading whitespaces would be eliminated by the `parseChunkExtensions()`. 

Feel free to refactor it or push any other changes.

**Comment by rousskov** [comment]
Please do not forget to push your changes. I will wait for them to show up before making mine. Thank you.

**Comment by uhliarik** [comment]
I found one more issue in my change after testing it. You can push it if you have something. Sorry for the fuzz.

**Comment by uhliarik** [comment]
I pushed everything what I have. This is the latest version which I tested and it worked for me. Feel free to do any changes you want. Thanks a lot for a review Alex.

**Review by rousskov — DISMISSED** [review]
@uhliarik, thanks, in part, to your rewrite, I have realized that we are not dealing with a new bug here. We are dealing with a regression of bug [4492](https://bugs.squid-cache.org/show_bug.cgi?id=4492)! This fact addresses PR description [XXX](https://github.com/squid-cache/squid/pull/1914#pullrequestreview-2361075943) (now gone), and makes it a lot easier to accept these changes. I also polished PR code a little (commit f837f5ff).

Please _test_ my changes (as of commit f79936a). If you are happy with them, and there are no objections from others, then I plan to clear this regression-fixing PR for merging.

**Review by rousskov — APPROVED** [review]

**Comment by rousskov** [comment]
Jenkins, it is OK to test this PR.

**Comment by notroj** [comment]
Thank you very much @uhliarik and @rousskov - I agree the comment in the httpd code is wrong here, I've updated it referencing this discussion. https://github.com/apache/httpd/commit/914eef0353c17501a4535c76e8aa8fe489359bce

File: src/http/one/TeChunkedParser.cc
Comment (yadij) [reply src/http/one/TeChunkedParser.cc:132]: The new `ParseBws()` logic (unintentionally?) adds a new difference from other servers ... support for sequences of multiple WSP characters. It would be best not to have that logic separated away from the case documentation. see my other comment which should resolve all this.

File: src/http/one/TeChunkedParser.cc
Comment (yadij) [inline src/http/one/TeChunkedParser.cc:130]: This being a very specific and CVE related case, I see much benefit in avoiding the new special `ParseBws()` logic.

Please instead prefer paranoia level of clarity:
```suggestion
        // Also, this is not BWS. It is explicit "MUST reject" violation of RFC 9112 HTTP/1.1.
        (void)tok.skipOne(CharacterSet::WSP);
```

**Review by yadij — CHANGES_REQUESTED** [review]
FTR; The removal of this tolerance was due to a specific [CVE issue](https://github.com/squid-cache/squid/security/advisories/GHSA-j83v-w3p4-5cqh).

 * NginX have the specific tolerance being proposed here (by my suggested update).
 * httpd apparently added tolerance to follow us in this PR.

Since you have testing, if you can please check and confirm if this is still an issue with current IBM_HTTP_Server releases. Document it as only being relevant to "IBM_HTTP_Server version X or older" would be perfect to aid future removal.

**Comment by yadij** [comment]
Marking for backport, and also our CVE SQUID-2023_1.patch (both v5 and v6) will need these changed appended. So **please** be minimal.

**Comment by notroj** [comment]
> Since you have testing, if you can please check and confirm if this is still an issue with current IBM_HTTP_Server releases. Document it as only being relevant to "IBM_HTTP_Server version X or older" would be perfect to aid future removal.

We (Red Hat) don't have access to that, @covener do you know if "spaces after chunk-size" this was something fixed in some past version of IHS?

File: src/http/one/TeChunkedParser.cc
Comment (uhliarik) [reply src/http/one/TeChunkedParser.cc:130]: Hello Amos. As Alex mentioned in a0d4fe17 , there has been seen two spaces after chunk-size. Suggested change  therfor would not fix such case.

**Comment by uhliarik** [comment]
@rousskov I tested the patch including all your commits in PR and it's working fine.

**Comment by covener** [comment]
> > Since you have testing, if you can please check and confirm if this is still an issue with current IBM_HTTP_Server releases. Document it as only being relevant to "IBM_HTTP_Server version X or older" would be perfect to aid future removal.
> 
> We (Red Hat) don't have access to that, @covener do you know if "spaces after chunk-size" this was something fixed in some past version of IHS?

I am the maintainer at IBM and have been working on it in some capacity since 2005.
It's nothing I recognize, and I don't see any such behavior with a basic CGI in our oldest in-service release.  It is not an area where we would intentionally deviate from upstream Apache HTTP Server.

It is possibly an Apache 1.3-based release, which would have been past any extended support 5 years prior to the 2016 bug.  

I'd suggest a commit message like "Likely affected historical IBM HTTP Server releases like 1.3.x. Contact IBM Support if you are adversely affected by this change."

File: src/http/one/TeChunkedParser.cc
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc:130]: In addition to what Luboš has [said](https://github.com/squid-cache/squid/pull/1914/files#r1796842949), skipOne() and skipAll() do not work well for incremental parsing of bad/prohibited-by-protocol input, as was already pointed out in the [initial review](https://github.com/squid-cache/squid/pull/1914#discussion_r1795888509) of this code. Please do not go back to the already rejected code.

File: src/http/one/TeChunkedParser.cc
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc:132]: > The new `ParseBws()` logic ...

This PR does not change ParseBws() logic.

> The new .... logic unintentionally?) adds a new difference from other servers ... support for sequences of multiple WSP characters.

That is not really _new_ logic: Squid already allowed multiple WSP characters after chunk-size before the regression (that this PR is fixing) was introduced. Please see [PR description](https://github.com/squid-cache/squid/pull/1914#issue-2579454036) for details.

File: src/http/one/TeChunkedParser.cc
Comment (yadij) [reply src/http/one/TeChunkedParser.cc:130]: Very well, skip twice then.

**Review by rousskov — COMMENTED** [review]
> [Amos](https://github.com/squid-cache/squid/pull/1914#pullrequestreview-2362277751): FTR; The removal of this tolerance was due to a specific [CVE issue](https://github.com/squid-cache/squid/security/advisories/GHSA-j83v-w3p4-5cqh).

While 2023 commit 951013d0 changes were triggered by that CVE, this PR does not reintroduce the two specific bugs that CVE was based on AFAICT.

> [Luboš](https://github.com/squid-cache/squid/pull/1914#issuecomment-2407249284): I tested the patch including all your commits in this PR and it's working fine.

Thank you very much. Unfortunately, this PR is now blocked by others. It cannot be merged until there are no negative votes. I tried to address Amos' change requests in this review. The ball is on @yadij side.

**Review by yadij — APPROVED** [review]

File: src/http/one/TeChunkedParser.cc
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc:130]: > Very well, skip twice then.

Skipping twice would not address my [concern](https://github.com/squid-cache/squid/pull/1914#discussion_r1796949281). Moreover, correctly skipping at most twice requires heroic efforts including creation of new parsing state. Simply calling skipOnce() twice would not accomplish that.

**Comment by notroj** [comment]
Thanks very much @covener. For completeness, in the traces we have from customers currently facing this issue there was no obvious presence of IBM HTTP Server, they all had `Server: Apache`. The only way this was associated with IHS I'm aware of is via https://bugs.squid-cache.org/show_bug.cgi?id=4492

**Pull request closed without merging**
